### PR TITLE
Extract task nugetSpec from nugetPack to generate nuspec

### PR DESCRIPTION
**Pull request by ngyukman**
On top of previous pull request

- support Map for defining nuspec in nugetSpec which could allow more flexible nuspec generation like doing modification before generation
- attemp to change the way on getting output based on task dependency, to avoid depending the output in configuration phase but could be changed during execution

### CODE REVIEW

File: src/main/groovy/com/ullink/NuGetPush.groovy [hunk 1]
```diff
@@ -11,19 +11,17 @@ class NuGetPush extends BaseNuGet {
...
-        project.afterEvaluate {
-            if (nupkgFile) {
-                def tasks = project.tasks.withType(NuGetPack.class).findAll { it.packageFile == nupkgFile }
-                if (tasks.size() == 1)
-                    dependsOn tasks[0]
-            }
-        }
+    }
+
+    String getNugetPackOutputFile() {
```
Comment (gluck) [inline src/main/groovy/com/ullink/NuGetPush.groovy]: this doesn't seem consistent with the dependsOn logic change.

On one hand you removed the discovery of the matching nugetPack task for the nupkg nugetPush wants to push (you put instead opinionated dependency for default tasks, which I'm fine with), but on the other hand nupkgFile is now *only* defaulted if there's a single nugetPack task.

File: src/main/groovy/com/ullink/NuGetPack.groovy [hunk 2]
```diff
@@ -69,16 +70,8 @@ class NuGetPack extends BaseNuGet {
...
-        nuspec = closure
-    }
-
-    Closure getNuspecCustom() {
-        nuspec
-    }
-
-    GPathResult getNuspec() {
-        new XmlSlurper().parse(getNuSpecFile())
+    Task getDependentNugetSpec() {
```
Comment (gluck) [inline src/main/groovy/com/ullink/NuGetPack.groovy]: return type can be NuGetSpec ?

File: src/main/groovy/com/ullink/NuGetPack.groovy [hunk 3]
```diff
@@ -86,14 +79,28 @@ class NuGetPack extends BaseNuGet {
...
+            return new XmlSlurper(false, false).parseText(nugetSpec.generateNuspec())
+        }
+    }
+
+    File getNuspecFile() {
+        if (nuspecFile) {
+            return project.file(this.nuspecFile)
+        }
+        def nugetSpec = getDependentNugetSpec()
+        if (nugetSpec && nugetSpec.nuspecFile.exists()) {
```
Comment (gluck) [inline src/main/groovy/com/ullink/NuGetPack.groovy]: why the exists check ? it'll only work if the method is called after the task is executed, and I don't think it's necessary

File: src/main/groovy/com/ullink/NuGetPack.groovy [hunk 4]
```diff
@@ -69,16 +70,8 @@ class NuGetPack extends BaseNuGet {
         super.exec()
     }
 
-    void nuspec(Closure closure) {
```
Comment (gluck) [inline src/main/groovy/com/ullink/NuGetPack.groovy]: any chance we can keep the backward compat by keeping this method and forwarding the setup to nugetSpec task ?

File: src/main/groovy/com/ullink/NuGetPush.groovy (hunk 1 above)
Comment (ngyukman) [inline src/main/groovy/com/ullink/NuGetPush.groovy]: Indeed I forgot this part after refactoring on the way working with dependsOn

File: src/main/groovy/com/ullink/NuGetPack.groovy (hunk 4 above)
Comment (ngyukman) [inline src/main/groovy/com/ullink/NuGetPack.groovy]: Sure, done

File: src/main/groovy/com/ullink/NuGetPack.groovy (hunk 3 above)
Comment (ngyukman) [inline src/main/groovy/com/ullink/NuGetPack.groovy]: it was due to getNuspec is called in project.afterEvaluate but the file is not there yet (unlike the previous implementation), moved to a more suitable place for the file exists check

File: src/main/groovy/com/ullink/NuGetPack.groovy (hunk 2 above)
Comment (ngyukman) [inline src/main/groovy/com/ullink/NuGetPack.groovy]: Done

File: README.md [hunk 5]
```diff
@@ -22,23 +26,24 @@ Sample usage:
...
-				authors 'Francois Valdy'
-				delegate.description '''some looong description...'''
-				// ...
-			}
-			delegate.files() {
-				delegate.file(src: 'somefile', target: 'tools')
-			}
-		}
+    nugetSpec {
+        // Closure will be applied to XMLBuilder
```
Comment (gluck) [inline README.md]: doc is now a bit outdated since the sample is using a map and not a closure

File: src/main/groovy/com/ullink/NuGetPush.groovy [hunk 6]
```diff
@@ -11,19 +11,20 @@ class NuGetPush extends BaseNuGet {
...
+        if (dependentNugetSpec)
+           dependentNugetSpec.packageFile
+    }
+
+    NuGetPack getDependentNugetSpec() {
```
Comment (gluck) [inline src/main/groovy/com/ullink/NuGetPush.groovy]: method name should be updated as well

Comment (ngyukman) [inline src/main/groovy/com/ullink/NuGetPush.groovy]: oh no :(

**Comment by gluck** [comment]
:shipit:

**Pull request merged**